pkg/util/limiter/query_limiter.go

```diff
@@ -58,6 +66,13 @@ type QueryLimiter struct {
 // NewQueryLimiter makes a new per-query limiter. Each query limiter is configured using the
 // `maxSeriesPerQuery`, `maxChunkBytesPerQuery`, and `maxChunksPerQuery` limits.
 func NewQueryLimiter(maxSeriesPerQuery, maxChunkBytesPerQuery, maxChunksPerQuery int, queryMetrics *stats.QueryMetrics) *QueryLimiter {
+	// Ensure the reject metric is initialised (so that we export the value "0" before a limit is reached for the first time).
+	if queryMetrics != nil {
+		for _, reason := range rejectReasons {
+			queryMetrics.QueriesRejectedTotal.WithLabelValues(reason)
+		}
+	}
```

Comment:
Rather than doing this here (and therefore for every query request), what if we did it once when the metrics are first initialised, in [`NewQueryMetrics`](https://github.com/grafana/mimir/blob/3d56c66c519d9b93da8ded458204cff702419ca7/pkg/querier/stats/query_metrics.go#L27)?

Reply:
Definitely. I should have thought about it. Done in https://github.com/grafana/mimir/pull/5440/commits/4f8eb921343c1569237da7eb45f33f02d4226ae0

Reply:
The downside of https://github.com/grafana/mimir/commit/4f8eb921343c1569237da7eb45f33f02d4226ae0 is that `cortex_querier_...` metric is then exposed by distributors too. We have a check in integration tests to ensure we only export right metrics from the right component, and these tests fail now.

I'm not sure how to fix it, while keeping the metrics initialisation in [NewQueryMetrics](https://github.com/grafana/mimir/blob/3d56c66c519d9b93da8ded458204cff702419ca7/pkg/querier/stats/query_metrics.go#L27).

Reverting https://github.com/grafana/mimir/commit/4f8eb921343c1569237da7eb45f33f02d4226ae0 for now.

Reply:
How about something like https://github.com/grafana/mimir/pull/5450? It's a much more involved change, but it also feels more clean.

Reply:
💯 it's way cleaner. Approved!